### PR TITLE
allow native and ssh-keygen public key check

### PR DESCRIPTION
**Pull request by Gibheer**
Hi,

this is the first part of fixing #2179, namely the SELinux and ssh-keygen problem. The first commit is pretty large, but adding public keys now works again using ssh-keygen.

Please do not merge yet, as there is still some stuff missing. I introduced new settings to make it possible to adjust the working directory for ssh-keygen and even adjust the ssh-keygen path. I have to update all occurrences of ssh-keygen for that still.
But if possible, I welcome everyone to test the patch :)

### CODE REVIEW

**Comment by Unknwon** [comment]
PS: build is failing...

**Comment by Gibheer** [comment]
Now this error I have no idea how to fix that. I tried with openssh 5.9 and that returns "is not a public key file" for ed25519 public keys. But what is travis using in this case, that it returns "file too long"?

I would like to keep the tests, but if it breaks the build, it is kind of stupid. Should I delete them or should I just remove the ed25519 key from the test?

**Comment by Unknwon** [comment]
I guess we need special instructions to tell Travis (but maybe not possible?) support ed25519 key.

Is this PR finished now?

**Comment by Gibheer** [comment]
Not yet. I want to adjust all occurences of ssh-keygen to use the new variables. But if you want, you could pull this request in and the other stuff will be done extra.

File: models/ssh_key.go [hunk 1]
```diff
@@ -33,7 +35,10 @@ const (
 	_TPL_PUBLICK_KEY = `command="%s serv key-%d --config='%s'",no-port-forwarding,no-X11-forwarding,no-agent-forwarding,no-pty %s` + "\n"
 )
 
-var sshOpLocker = sync.Mutex{}
+var (
+	sshOpLocker          = sync.Mutex{}
+	SSH_UNKNOWN_KEY_TYPE = fmt.Errorf("unknown key type")
```
Comment (Unknwon) [inline models/ssh_key.go]: Variable should not use ALL_CAPS.

**Comment by Unknwon** [comment]
What is the different between this PR and revert old approach back?

**Comment by Gibheer** [comment]
This is a combination of the old behavior and the new one. At the end it should be possible to use the go native ssh library only without any dependency on ssh-keygen at all. Currently, there are still some places in need of ssh-keygen.
That would also mean, that depending on what server you choose to use (sshd, gogs), you would get the full support from the specific system.

File: modules/setting/setting.go [hunk 2]
```diff
@@ -30,9 +30,11 @@ import (
 type Scheme string
 
 const (
-	HTTP  Scheme = "http"
-	HTTPS Scheme = "https"
-	FCGI  Scheme = "fcgi"
+	HTTP                       Scheme = "http"
+	HTTPS                      Scheme = "https"
+	FCGI                       Scheme = "fcgi"
+	SSH_PUBLICKEY_CHECK_NATIVE        = "native"
```
Comment (Unknwon) [inline modules/setting/setting.go]: This `const` section is for `Scheme`, do not mixed up with them.

File: modules/setting/setting.go [hunk 3]
```diff
@@ -328,6 +333,29 @@ func NewContext() {
 	if err := os.MkdirAll(SSHRootPath, 0700); err != nil {
 		log.Fatal(4, "Fail to create '%s': %v", SSHRootPath, err)
 	}
+	checkDefault := SSH_PUBLICKEY_CHECK_KEYGEN
+	if DisableSSH {
+		checkDefault = SSH_PUBLICKEY_CHECK_NATIVE
+	}
```
Comment (Unknwon) [inline modules/setting/setting.go]: When disable SSH, why should we care the method at all?

**Comment by Unknwon** [comment]
ping @Gibheer

**Comment by Gibheer** [comment]
@Unknwon sorry, took a while longer. I fixed the stuff you found and replaced the ssh-keygen call with the new variable from setting. You can merge this for now, as I do not know when I have the time to do the rest (making ssh-keygen only necessary when the system sshd is used).
Thank you for your help :)

**Comment by Unknwon** [comment]
Thanks!

**Pull request merged**